File: control/src/cpp_controller/src/vehicle_pid_controller.cpp
#include <cpp_controller/vehicle_pid_controller.hpp>


// 横向PID控制器
PIDLateralController::PIDLateralController(double K_PP, double C_PP, double I_PP, double D_PP): 
            _K_PP(K_PP) , _C_PP(C_PP) , _I_PP(I_PP) , _D_PP(D_PP) {}

// 计算两向量的夹角
double PIDLateralController::calculate_angle(const std::array<double, 2>& v1, const std::array<double, 2>& v2) {
    double dot_product = v1[0] * v2[0] + v1[1] * v2[1];
    double magnitude_v1 = std::hypot(v1[0], v1[1]);
    double magnitude_v2 = std::hypot(v2[0], v2[1]);
    double cos_theta = dot_product / (magnitude_v1 * magnitude_v2);
    return std::acos(cos_theta);
}

// 计算两点之间的距离
double PIDLateralController::calculate_distance(const std::array<double, 2>& v1, const std::array<double, 2>& v2) {
    return std::hypot(v1[0] - v2[0], v1[1] - v2[1]);
}

// run_step 方法实现
double PIDLateralController::run_step(
    const std::array<double, 3>& current_pose, 
    const std::deque<std::array<double, 2>> _waypoints_queue,
    const double current_speed) 
{
    double yaw = current_pose[2];
    std::array<double, 2> v_begin = {current_pose[0], current_pose[1]};
    std::array<double, 2> v_end = {v_begin[0] + std::cos(yaw), v_begin[1] + std::sin(yaw)};
    std::array<double, 2> v_vec = {v_end[0] - v_begin[0], v_end[1] - v_begin[1]};  // 当前航向向量

    lookahead_distance = _K_PP * current_speed + _C_PP;

    // 先找出前视点，再计算目标航向向量
    
    // 第一版：直接找第一个大于等于前视距离的点（默认路径队列从近到远排列）
    for(auto& waypoints:_waypoints_queue){
        waypoint = waypoints;
        if (calculate_distance(v_begin, waypoints) >= lookahead_distance) {
            break;
        }
    }
    

    // 第二版：找出距离最近的点
    /*
    double min_distance = 100000.0;
    bool found = false;
    for(auto& waypoints:_waypoints_queue){
        double temp_distance = calculate_distance(v_begin, waypoints);
        if (temp_distance >= lookahead_distance && temp_distance < min_distance) {
            min_distance = temp_distance;
            waypoint = waypoints;
            found = true;
        }
    }
    if (!found) {
        waypoint = _waypoints_queue.back();
    }
    */

    std::array<double, 2> w_vec = {waypoint[0] - v_begin[0], waypoint[1] - v_begin[1]};  // 目标航向向量

    // 计算两向量的夹角
    double _dot = calculate_angle(v_vec, w_vec);

    // 计算叉乘，判断转向方向
    double _cross = v_vec[0] * w_vec[1] - v_vec[1] * w_vec[0];
    if (_cross < 0) {
        _dot *= -1.0;
    
    }

    

    double previous_error = error;
    error = lookahead_distance * std::sin(_dot);

    // 限制积分项，避免积分饱和
    error_integral = std::clamp(error_integral + error, -40.0, 40.0);
    error_derivative = error - previous_error;

    // 计算PID输出
    double output = atan(2 * wheel_base * error / pow(lookahead_distance,2))+ _I_PP * error_integral + _D_PP * error_derivative;
    double steer = std::clamp(output/0.380635, -1.0, 1.0);  // 限制输出范围
    //double steer = std::clamp(output, -1.0, 1.0);  // 限制输出范围

    // 如果变化过快，保持上一次的转向值
    if (std::abs(steer - last_steer) > 1.0) {
        return last_steer;
    } 
    else {       
        last_steer = steer;
        return steer;
    }
}

// 纵向PID控制器
PIDLongitudinalController::PIDLongitudinalController(double K_P, double K_D, double K_I)
    : _K_P(K_P), _K_D(K_D), _K_I(K_I), error(0.0), error_integral(0.0), error_derivative(0.0) {}

// run_step 方法实现
double PIDLongitudinalController::run_step(double target_speed, double current_speed) {
    double previous_error = error;
    error = target_speed - current_speed;

    // 限制积分项，避免积分饱和
    error_integral = std::clamp(error_integral + error, -40.0, 40.0);
    error_derivative = error - previous_error;

    // 计算PID输出
    double output = _K_P * error + _K_I * error_integral + _K_D * error_derivative;

    // 限制输出范围
    return std::clamp(output, -1.0, 1.0);
}

// 车辆PID控制器
VehiclePIDController::VehiclePIDController(const std::map<std::string, double>& args_lateral,
                                           const std::map<std::string, double>& args_longitudinal)
    : _lon_controller(args_longitudinal.at("K_P"), args_longitudinal.at("K_D"), args_longitudinal.at("K_I")),
      _lat_controller(args_lateral.at("K_PP"), args_lateral.at("C_PP"), args_lateral.at("I_PP"), args_lateral.at("D_PP")) {}

// run_step 方法实现
std::pair<double, double> VehiclePIDController::run_step(
    double target_speed, 
    double current_speed, 
    const std::array<double, 3>& current_pose, 
    const std::deque<std::array<double, 2>> _waypoints_queue) 
{
    // 调用纵向控制器计算油门值
    double throttle = _lon_controller.run_step(target_speed, current_speed);

    // 调用横向控制器计算转向值
    double steering = _lat_controller.run_step(current_pose, _waypoints_queue, current_speed);

    // 返回转向和油门值
    return {steering, throttle};
}

File: control/src/simple_pid_controller/simple_pid_controller/vehicle_pid_controller.py
from collections import deque  # 用于创建双端队列
import math
import numpy as np


class ControllerCFG:
    def __init__(self):
        # PID控制器参数
        self.lateral_KP = 0.9
        self.lateral_KI = 0.0
        self.lateral_KD = 0.0
        self.longitudinal_KP = 0.05
        self.longitudinal_KI = 0.0
        self.longitudinal_KD = 0.0
        # 限制
        self.max_angle_error = 100.0
        self.max_angle_derivative = 150.0
        self.max_steer_diff = 1.5
        self.recovery_lateral_KP = 0.6
        self.recovery_lateral_KD = 0.1
        self.recovery_longitudinal_KP = 0.1 


class VehiclePIDController(object):  # pylint: disable=too-few-public-methods
    """
    VehiclePIDController is the combination of two PID controllers (lateral and longitudinal)
    to perform the low level control a vehicle from client side
    """
    def __init__(self):
        self.cfg = ControllerCFG()
        self._lon_controller = PIDLongitudinalController()  # 纵向PID控制器
        self._lat_controller = PIDLateralController()  # 横向PID控制器

    def run_step(self, target_speed, current_speed, current_pose, waypoint):
        """
        Execute one step of control invoking both lateral and longitudinal
        PID controllers to reach a target waypoint at a given target_speed.
        """
        throttle = self._lon_controller.run_step(target_speed, current_speed)  # 油门控制
        steering = self._lat_controller.run_step(current_pose, waypoint)  # 转向控制

        return steering, throttle


class PIDLongitudinalController(object):  # pylint: disable=too-few-public-methods
    """
    PIDLongitudinalController implements longitudinal control using a PID.
    """
    def __init__(self):
        self.cfg = ControllerCFG()
        self.error = 0.0  # 误差
        self.error_integral = 0.0  # 积分项误差
        self.error_derivative = 0.0  # 微分项误差

    def run_step(self, target_speed, current_speed):
        """
        Estimate the throttle of the vehicle based on the PID equations
        """
        previous_error = self.error
        self.error = target_speed - current_speed
        # restrict integral term to avoid integral windup
        self.error_integral = np.clip(self.error_integral + self.error, -40.0, 40.0)
        self.error_derivative = self.error - previous_error
        output = (self.cfg.longitudinal_KP * self.error +
                  self.cfg.longitudinal_KI * self.error_integral +
                  self.cfg.longitudinal_KD * self.error_derivative)
        return np.clip(output, -1.0, 1.0)


class PIDLateralController(object):  # pylint: disable=too-few-public-methods
    """
    PIDLateralController implements lateral control using a PID.
    """
    def __init__(self):
        self.cfg = ControllerCFG()
        self._e_buffer = deque(maxlen=10)
        self.error = 0.0
        self.error_integral = 0.0
        self.error_derivative = 0.0
        self.last_output = 0.0
        self.last_steer = 0.0

    def run_step(self, current_pose, waypoint):
        
        # if self._in_recovery_mode:  # 需要增加状态传递
        #     kp = self.cfg.recovery_lateral_KP
        #     kd = self.cfg.recovery_lateral_KD
        # else:
        #     kp = self.cfg.lateral_KP
        #     kd = self.cfg.lateral_KD
        """
        Estimate the steering angle of the vehicle based on the PID equations
        """
        v_begin = np.array(current_pose[:2])
        yaw = current_pose[2]
        v_end = np.array([v_begin[0] + math.cos(yaw), v_begin[1] + math.sin(yaw)])

        v_vec = v_end - v_begin  # 当前航向向量
        w_vec = np.array(waypoint) - v_begin  # 目标航向向量
        _dot = math.acos(np.clip(np.dot(w_vec, v_vec) /
                                 (np.linalg.norm(w_vec) * np.linalg.norm(v_vec)), -1.0, 1.0))  # 计算两向量夹角

        _cross = np.cross(v_vec, w_vec)  # 计算两向量叉乘，用于判断转向方向
        if _cross < 0:
            _dot *= -1.0

        previous_error = self.error
        present_error = _dot
        # restrict integral term to avoid integral windup
        error_integral = np.clip(self.error_integral + present_error, -400.0, 400.0)
        error_derivative = present_error - previous_error
        output = (self.cfg.lateral_KP * present_error +
                  self.cfg.lateral_KI * error_integral +
                  self.cfg.lateral_KD * error_derivative)
        steer = np.clip(output, -1.0, 1.0)
        if (abs(present_error) > math.radians(self.cfg.max_angle_error) or
                abs(error_derivative) > math.radians(self.cfg.max_angle_derivative) or
                abs(steer - self.last_steer) > self.cfg.max_steer_diff):
            return self.last_steer
        else:
            self.error = present_error
            self.error_integral = error_integral
            self.error_derivative = error_derivative
            self.last_output = output
            self.last_steer = steer
            return steer
